Feat/odometry

dorukayhan opened the pull request:
It builds, runs, and seems to return numbers that make sense without crashing when I poke it in rqt. More meaningful testing could be done in the future when we have a robot.

dorukayhan committed:
particlefilter.py from last year

also added commentary for future us, also turned a few lists that should've
been tuples into tuples (a list is arbitrarily many things of a common type;
what we want is a single thing with components, which a tuple fits better)

dorukayhan committed:
incorporate fixthecore's work (1b8b6b8f, 2c1ddaab)

node.py changes seem important
not like the commit graph can get much uglier, me install update

dorukayhan committed:
filters.py from last year

ok i think i understand the architecture now?
danger zone does odometry in the autonav_filters node, and it's structured
as a main script that runs the show and one or more distinct filters (dead
reckoning and particle filter). also there's c++ for some reason and it
seems less organized than and unrelated to the python stuff. somehow. this
autonav_odometry could be similar but python only and with just a particle
filter to start

dorukayhan committed:
clean up main.py

update type names, remove dead reckoning, commentate, may need to change the
filter terminology too

dorukayhan committed:
having a working build is important innit?

dorukayhan committed:
we can now teleport across state lines without crashing particlefilter

also it has debug output now
comes in very handy when testing the odometer in rqt with messages that make
no temporal sense (e.g. latitude changes by 1 deg between two gps readings)

danielbrownmsm commented on autonav_ws/src/autonav_odometry/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+# find dependencies
+find_package(ament_cmake_python REQUIRED)
+find_package(rclpy REQUIRED)
+find_package(autonav_shared REQUIRED)
```
you might need a `find_package(autonav_msgs REQUIRED)` in here too, I think, can't remember

danielbrownmsm commented on autonav_ws/src/autonav_odometry/package.xml:
```diff
@@ -0,0 +1,18 @@
+<?xml-model href="http://download.ros.org/schema/package_format3.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="3">
+    <name>autonav_odometry</name>
+    <version>0.1.0</version>
```
2025.0.1 (or 2025.1.0, whichever you prefer) please

antonioc76 replied:
2025.x.y, follow example in autonav_example_py

danielbrownmsm commented on autonav_ws/src/autonav_odometry/package.xml:
```diff
@@ -0,0 +1,18 @@
+    <buildtool_depend>ament_cmake_python</buildtool_depend>
+
+    <depend>rclpy</depend>
+    <depend>autonav_shared</depend>
```
might need a `<depend>autonav_msgs</depend>` here, can't remember

danielbrownmsm commented on autonav_ws/src/autonav_odometry/resource/autonav_odometry:
empty file?

danielbrownmsm commented on autonav_ws/src/autonav_odometry/src/main.py:
```diff
@@ -0,0 +1,113 @@
+from autonav_msgs.msg import MotorFeedback, GPSFeedback, Position
+from autonav_shared.types import DeviceState, SystemState, LogLevel # i am goig to beat pylance with a rock
+from particlefilter import ParticleFilter
+# from scr_msgs.msg import SystemState
```
this was from last year's Core, you can safely delete this line

danielbrownmsm commented on autonav_ws/src/autonav_odometry/src/main.py:
```diff
@@ -0,0 +1,113 @@
+import json
+from types import SimpleNamespace
+from autonav_msgs.msg import MotorFeedback, GPSFeedback, Position
+from autonav_shared.types import DeviceState, SystemState, LogLevel # i am goig to beat pylance with a rock
```
*going

danielbrownmsm commented on autonav_ws/src/autonav_odometry/src/main.py:
```diff
@@ -0,0 +1,113 @@
+        self.config = FiltersNodeConfig()
+        self.onReset()
+
+    def config_updated(self, jsonObject):
```
this is from last year's config system, you can safely remove it

danielbrownmsm commented on autonav_ws/src/autonav_odometry/src/main.py:
```diff
@@ -0,0 +1,113 @@
+    def init(self):
+        # [melee announcer voice] ready...
+        self.create_subscription(GPSFeedback, "/autonav/gps", self.onGPSReceived, 20)
+        self.create_subscription(MotorFeedback, "/autonav/MotorFeedback", self.onMotorFeedbackReceived, 20)
```
should be `"/autonav/motor_feedback"`

danielbrownmsm commented on autonav_ws/src/autonav_odometry/src/main.py:
```diff
@@ -0,0 +1,113 @@
+            self.onReset()
+
+    def onGPSReceived(self, msg: GPSFeedback):
+        if msg.gps_fix == 0 and msg.is_locked == False:
```
`msg.is_locked` does not exist anymore (see https://github.com/SoonerRobotics/autonav_software_2025/blob/feat/vectornav/autonav_ws/src/autonav_msgs/msg/GPSFeedback.msg)
just use msg.gps_fix. I can get you the vectornav PDF documentation on what the gps_fix number actually means (as I don't remember it at the moment)

danielbrownmsm commented on autonav_ws/src/autonav_odometry/src/main.py:
```diff
@@ -0,0 +1,113 @@
+        position = Position()
+        position.x = averages[0]
+        position.y = averages[1]
+        position.theta = averages[2]
```
if you wanted to you could do `position.x, position.y, position.theta = averages` but I don't really care either way, just plugging some Python tricks

danielbrownmsm commented on autonav_ws/src/autonav_odometry/src/particlefilter.py:
```diff
@@ -0,0 +1,102 @@
+        self.gps_noise = 0.45
+        self.odom_noise = (0, 0, 0.1)
+        # TODO should these parameters be in a config file in autonav_odometry/resource?
+        self.init_particles()
```
bit of a redundant call (as you call onReset() in __init__ for FiltersNode line 34) but it's not that big of a deal.
what's a bigger deal is using the seed heading, which I think. . . actually, nevermind, because init_particles is called in FiltersNode after initialization when we would actually be able to send it a seed heading, so that's a problem for FiltersNode not ParticleFilter

antonioc76 replied:
noise values should absolutely be in the config

danielbrownmsm commented on autonav_ws/src/autonav_odometry/src/particlefilter.py:
```diff
@@ -0,0 +1,102 @@
+        for particle in self.particles:
+            # update particles
+            # MotorFeedback gives robot-relative data (x is forward, y is left (???)) so we convert it to track-relative with trig magic  
+            # TODO why are we scaling down delta_x and delta_theta again?
```
that would be a Tony question but basically because magic number
we might not need it with the swerve drive (or we might need a different magic number), that remains to be seen

antonioc76 replied:
These are fudge factors and they should probably go in the config. Last year I used 1.2 for example. Just depends on the robot/ simulator.

danielbrownmsm commented on autonav_ws/src/autonav_odometry/src/main.py:
```diff
@@ -0,0 +1,113 @@
+    def onReset(self):
+        self.first_gps = None
+        self.last_gps = None
+        self.pf.init_particles()
```
I think we should use the seed heading parameter of init_particles, or at least make whether we use it or not controlled by the config. this would've been handy last year, as we only ever started facing one direction, whereas in 2023 you could start facing north or south.

danielbrownmsm commented on autonav_ws/src/autonav_odometry/src/main.py:
```diff
@@ -0,0 +1,113 @@
+        self.pf.init_particles()
+        # INIT NEW FILTERS HERE
+
+    def system_state_transition(self, old: SystemState, updated: SystemState):
```
doesn't currently exist this year in the Core but this is probably a good argument for implementing it this year so leave this here for now, and if it breaks when building because this function doesn't exist/get called then nudge Dylan or I and we can add it

danielbrownmsm commented on autonav_ws/src/autonav_odometry/src/main.py:
```diff
@@ -0,0 +1,113 @@
+        self.first_gps = None
+        self.last_gps = None
+        self.latitude_length = self.declare_parameter("latitude_length", 111086.2).get_parameter_value().double_value
+        self.longitude_length = self.declare_parameter("longitude_length", 81978.2).get_parameter_value().double_value
```
this might be more of an in-person conversation (maybe probably with Tony) but I think lat and long length should be part of the config instead of being launch file parameters.

danielbrownmsm reviewed (commented):
just a few notes and things, if you have any questions feel free to ask (and Tony is probably the better person to ask because he wrote this, I don't know how the particle filter actually works.

antonioc76 commented on autonav_ws/src/autonav_odometry/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+install(PROGRAMS
+  # enumerate every file in src by hand here like a bozo because globs don't just work i guess
+  src/main.py
+  src/particlefilter.py
```
particlefilter.py just creates a class that the node uses. It doesn't need to be enumerated here since it's not a node.

antonioc76 reviewed (changes requested):
Good way to test is to just instantiate a particlefilter object and run it on last years data, which is all time stamped. Data is available in autonav_software_2024 on the pftest-tony branch, in autonav_filters/tests. 

I believe ros2 has support for pytest with colcon test. I think it would be great for the team to learn how to use that and round out our testing framework knowledge.

dorukayhan committed:
catch up with three months of work

everything from 2c1ddaab to ff53c98b. none of it is odometry, which means
i get to not discard anything, yay
